add font8x8 rbg565 support for cfb on rt1050

hakehuang opened the pull request:
enable customer font test in font 8x8
enable cfb, cfb shell with font 8x8

Signed-off-by: Hake Huang <hake.huang@nxp.com>

microbuilder commented on samples/display/cfb_custom_font/CMakeLists.txt:
```diff
@@ -9,4 +9,8 @@ project(cfb_custom_font)
 
 set(gen_dir ${ZEPHYR_BINARY_DIR}/include/generated/)
+if(CONFIG_DISPLAY_MCUX_ELCDIF)
+generate_cfb_font_for_target(app fonts/8x8text_darkGrayNoShadow.png ${gen_dir}/cfb_font_dice.h 8 8 --first 49 --last 54)
```
While the mono version of this has previously been merged in, extending this example to include basic latin characters would perhaps make this more useful, rather than the current range of `49-54`, corresponding to `1-6` on the ASCII table. There is a need for small fonts with clear license terms, so having a clearly licensed 5x7 or 8x8 fixed-width font in RGB565 and monochrome would be useful here.

[Current font options](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/fb/cfb_fonts.c) are on the large side at 10x16, 15x24 and 20x32, so having an 8x8 version in this demo would address this issue.

hakehuang replied:
the font is edited from https://opengameart.org/content/8x8-font, as the customer font demos 1-6, so I only add 1-6 in the png

microbuilder replied:
NP. I'll try to get a larger character in as a separate PR using this method, as well as finding an appropriately licensed 5x7 font.

hakehuang replied:
I add subsys/fb/font_8x8.c from Linux kernel fonts, which is a full set

vanwinkeljan commented on subsys/fb/cfb.c:
```diff
@@ -116,16 +164,40 @@ int cfb_print(struct device *dev, char *str, u16_t x, u16_t y)
+		}
+		break;
+	case PIXEL_FORMAT_RGB_565:
 		for (size_t i = 0; i < strlen(str); i++) {
```
Looks like a lot of code is duplicated here. Why not just select the correct draw_char function and keep the for loop common?

hakehuang replied:
right, will do that

vanwinkeljan commented on subsys/fb/font_8x8.c:
```diff
@@ -0,0 +1,2586 @@
+﻿/**********************************************/
```
Missing license header

hakehuang replied:
this font is modified from linux kernel source, and there is not lincense in it see https://github.com/steev/linux-kernel/blob/master/drivers/video/console/font_8x8.c

vanwinkeljan replied:
Is the linux kernel not GPL licensed maybe better look for a font with a less strict license?

Please have a look at https://docs.zephyrproject.org/latest/contribute/index.html#contributing-non-apache-2-0-licensed-components

hakehuang replied:
I update with the OFL 1.1 license font download from the https://www.fontsquirrel.com/fonts/download/abeezee

vanwinkeljan reviewed (changes requested):
See comments and also address codeowner and git lint issue

hakehuang commented on samples/display/cfb/prj_mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_STDOUT_CONSOLE=y
```
ok

hakehuang commented on samples/display/cfb/src/main.c:
```diff
@@ -38,7 +38,7 @@ void main(void)
 	if (display_set_pixel_format(dev, PIXEL_FORMAT_MONO10) != 0) {
 		printf("Failed to set required pixel format\n");
-		return;
+		/* return;*/
```
the problem is mcux_elcd driver this pixel_format is not support, and for rgb565 panel it seems this is not a must.

vanwinkeljan replied:
Actually  due to adding RGB565 to CFB it works, looks to me that this code is just a safety check.
You could first do a call to `display_get_capabilities` to get the supported pixel formats and then set the one you need.

hakehuang replied:
ok

vanwinkeljan replied:
But with the proposed change it will always print a failure because it is setting a none supported mode and actually the code is setting the wrong mode.

So why not query the modes supported by the display (via `display_get_capabilities`) and apply one that is supported by CFB and if no supported mode is found report the error

hakehuang replied:
ok, let me do this way.

hakehuang commented on samples/display/cfb_custom_font/prj_mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_HEAP_MEM_POOL_SIZE=300000
```
ok

hakehuang commented on samples/display/cfb_shell/prj_mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_BOOT_BANNER=n
```
ok

hakehuang commented on subsys/fb/Kconfig:
```diff
@@ -15,7 +15,14 @@ if CHARACTER_FRAMEBUFFER
 config CHARACTER_FRAMEBUFFER_USE_DEFAULT_FONTS
 	bool "Use default fonts"
-	default y
+	default y if SSD1673
```
I do not full understand the default font, seems it is not that strieght forward as the font8x8, let me double check, maybe I shall add a convert function

hakehuang commented on subsys/fb/Kconfig:
```diff
@@ -15,7 +15,14 @@ if CHARACTER_FRAMEBUFFER
+	help
+	  Use default fonts.
+
+config CHARACTER_FRAMEBUFFER_USE_LINUX_FONTS
```
this font is  copied from linux kernel, will this be a license issue, if yes, maybe I need remove this.

hakehuang commented on subsys/fb/Kconfig:
```diff
@@ -31,6 +38,7 @@ if CHARACTER_FRAMEBUFFER_SHELL
 	string
 	default "SSD1673" if SSD1673
 	default "SSD1306" if SSD1306
+	default "ELCDIF_1" if DISPLAY_MCUX_ELCDIF
```
SSD1306 and SSD1673 is also put here. I just follow the default ones.

vanwinkeljan replied:
Reasoning behind the question is that board Kconfig.defconfig already contains one for LVGL and I thought it is better to group them in one place

hakehuang replied:
so move the  CHARACTER_FRAMEBUFFER_SHELL to board? this may need to change other platform which may not owned by me. can we just keep it as is for this time of being? @vanwinkeljan

vanwinkeljan replied:
I dont see how moving a config in to a board specific defconfig file would impact another platform.
The same has been done for the LVGL config for the same board:
https://github.com/zephyrproject-rtos/zephyr/blob/e0943cfaab29100c5212f74aa9f3eb1a4aa58f0b/boards/arm/mimxrt1050_evk/Kconfig.defconfig#L76

hakehuang replied:
LVGL_DISPLAY_DEV_NAME  in board default config, as the lvgl sample needs a display name, so maureen name it LVGL_DISPLAY_DEV_NAME, but in cfb cases the original code using the DISPLAY_NAME to refer to the esiplay driver. So if we keep adding defines in this way, the default board config will be too big. So the root cause is that every application developer create their own display naming, for cfb_shell it is  CHARACTER_FRAMEBUFFER_SHELL, for cfb it is the DISPLAY_NAME, and for lvgl it is LVGL_DISPLAY_DEV_NAME. I guess we can keep the rules as is before we can get aligned, any commnets? @vanwinkeljan

vanwinkeljan replied:
> So if we keep adding defines in this way, the default board config will be too big.

Actually  the defines could go into the board specific conf files of the samples.
So maybe better move it over there if the concern is that the board default Kconfig will grow to large.

>So the root cause is that every application developer create their own display naming

Main reason for this is the you could have multiple displays and both LVGL and CFB shell init run during system startup.

hakehuang replied:
ok， let me create the board config for each samples

hakehuang commented on subsys/fb/cfb.c:
```diff
@@ -66,6 +66,14 @@ static inline u8_t *get_glyph_ptr(const struct cfb_font *fptr, char c)
 	return NULL;
 }
 
+static inline u8_t *get_glyph_font8x8_ptr(const struct cfb_font *fptr, char c)
```
because I add a new font, which is different from the default one. let me check whether I can use the default one directly.

jfischer-no reviewed (dismissed):
* please split you commits in related parts (changes in subsys, changes in drivers, changes in samples ...)
* use font under Apache License, e.g. [DroidSansMono](https://www.fontsquirrel.com/fonts/Droid-Sans-Mono)
* use gen_cfb_font_header.py to generate custom font or new font subset for `subsys/fb/cfb_fonts.c`
e.g. `./scripts/gen_cfb_font_header.py -i DroidSansMono.ttf -t font -s 8 -x 8 -y 8 -o test.c --first 32 --last 127`

possibly, gen_cfb_font_header.py needs to be tuned to generate such small font.

hakehuang commented:
@jfischer-phytec-iot DroidSansMono is ugly in 16 bit panel. as the OFL font is not allow, I will try to find Apache license font.

jfischer-phytec-iot commented:
@hakehuang 
>DroidSansMono is ugly in 16 bit panel.

Can you please explain what is ugly? How ugliness in this case depends on panel type or pixel format?

>as the OFL font is not allow, I will try to find Apache license font.

I do not mean it is not allowed, you have to find out whether OFL and derived work are compatible with Apache License.

hakehuang commented:
@jfischer-phytec-iot  I try to convert the DroidSansMono font to 8x8 format, and it looks not good, for example for '2' the font matrix is like something '=', I find another font with Apache license, please check

vanwinkeljan commented on subsys/fb/Kconfig:
```diff
@@ -13,12 +13,26 @@ menuconfig CHARACTER_FRAMEBUFFER
+
+choice
+	prompt "Fonts in use"
+	default CHARACTER_FRAMEBUFFER_USE_DEFAULT_FONTS if SSD1673 || SSD1306
```
Is it  not an option to always select the default font and where need select any other supported font through Kconfig.

hakehuang replied:
@vanwinkeljan , 
![image](https://user-images.githubusercontent.com/800198/59992788-1394a600-9680-11e9-97f5-48cfe79af03b.png), this option is prompt a default font option and use can choose other existing fonts manually, currently there only two fonts supported. Can you give me more hints on how to change?

vanwinkeljan replied:
Just keep a single default (```default CHARACTER_FRAMEBUFFER_USE_DEFAULT_FONTS```) and dont check for display type

hakehuang replied:
@vanwinkeljan even the fonts are not applicable to default display panel?

vanwinkeljan replied:
It just does not feel right that for every new display we add we need to update the CFB Kconfig.
I would assume the default font is the default font for all displays and if another font is need it needs to be put in an overlay

hakehuang replied:
The situation is that initial font is a tiled one and the algorithm to display it is quite different from a plain text normal font table generated. I think the difference is only for the panel type, so there will not be many fonts added. as far as I know only tiled font and plain font.

vanwinkeljan replied:
Problem is not that there will be many fonts but a number of growing display controller drivers.
Dont know how to solve this in Kconfig (@SebastianBoe / @galak any suggestions)

Also it is now very confusing you have a Kconfig called `CHARACTER_FRAMEBUFFER_USE_DEFAULT_FONTS` which is catulay selects a specific font.
Would it not be better to provide 3 options in the choice:
 * default tiled font
 * default font
 * custom font

hakehuang replied:
ok, let me change to default
* default tiled font
* default font
* custom font

vanwinkeljan commented on subsys/fb/cfb.c:
```diff
@@ -310,9 +368,26 @@ int cfb_framebuffer_init(struct device *dev)
+	}
 	fb->buf = k_malloc(fb->size);
 	if (!fb->buf) {
+		LOG_DBG("malloc %d fails", fb->size);
```
Maybe log as error

hakehuang replied:
right, this is better, updated

hakehuang commented:
> See comments and also address codeowner and git lint issue

updated

vanwinkeljan commented on subsys/fb/cfb_shell.c:
```diff
@@ -19,8 +19,11 @@
 #define HELP_INIT "call \"cfb init\" first"
 #define HELP_PRINT "<col: pos> <row: pos> \"<text>\""
 
+#ifdef CONFIG_CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME
```
Either remove `CONFIG_CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME` completely or revert the changes and update the board specific configs i the samples.

hakehuang replied:
@vanwinkeljan , remove the CONFIG_CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME seems the only option, as it is a choice option, which can not be put into board specific configs in the samples.

vanwinkeljan replied:
As far as I can see `CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME` is just a string config and not part of a choice. Could it be that you update all CFB samples and not only the shell one?

hakehuang replied:
@vanwinkeljan , only shell example using this way, all other cfb examples use the DISPLAY_DRIVER, I do not know why it is originally designed so. So you think it is OK for me to change it? let me have a try.

vanwinkeljan replied:
Yes it forces the user to specify the display name to use for the shell. So can this change be reverted?

hakehuang replied:
yes

vanwinkeljan commented on subsys/fb/Kconfig:
```diff
@@ -31,6 +50,8 @@ if CHARACTER_FRAMEBUFFER_SHELL
 	string
 	default "SSD1673" if SSD1673
 	default "SSD1306" if SSD1306
+	default "TBD"
```
No need for a default value

hakehuang replied:
ok

vanwinkeljan commented on subsys/fb/Kconfig:
```diff
@@ -31,6 +50,8 @@ if CHARACTER_FRAMEBUFFER_SHELL
 	default "SSD1673" if SSD1673
 	default "SSD1306" if SSD1306
+	default "TBD"
+	depends on CHARACTER_FRAMEBUFFER_SHELL && CHARACTER_FRAMEBUFFER
```
No need for `depends on` as the config is already in identical if statements

hakehuang replied:
ok

vanwinkeljan commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -71,8 +71,18 @@ config ETH_MCUX_0
 endif # NETWORKING
 
+config DISPLAY_DRIVER
+	string
```
defconfigs should name a type, please remove `string` from both `DISPLAY_DRIVER`  and `CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME`.

hakehuang replied:
ok

vanwinkeljan commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -71,8 +71,18 @@ config ETH_MCUX_0
 
 endif # NETWORKING
 
+config DISPLAY_DRIVER
```
Should this config not be defined in the CFB KConfig and better rename it to `CHARACTER_FRAMEBUFFER__DISPLAY_DEV_NAME`?
Further, remove `CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME` and rework the samples and cf_shell.c to use this new config.

hakehuang replied:
ok

vanwinkeljan commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -71,8 +71,18 @@ config ETH_MCUX_0
 
 endif # NETWORKING
 
+config DISPLAY_DRIVER
```
I think both `DISPLAY_DRIVER` and `CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME` should be put under an if statement.

hakehuang replied:
ok

vanwinkeljan commented on subsys/fb/Kconfig:
```diff
@@ -27,7 +45,7 @@ config CHARACTER_FRAMEBUFFER_SHELL
 
 if CHARACTER_FRAMEBUFFER_SHELL
-	config CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME
+	config CHARACTER_FRAMEBUFFER__DISPLAY_DEV_NAME
```
nitpick remove the redundant `_` in front of `DISPLAY_DEV_NAME`

hakehuang replied:
ok

vanwinkeljan commented:
@hakehuang in general the PR looks good to me can you fix the remaining Shippable failure

vanwinkeljan commented on boards/arm/reel_board/Kconfig.defconfig:
```diff
@@ -81,4 +81,11 @@ config BT_CTLR
+if CHARACTER_FRAMEBUFFER
+
+config CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME
+	default "SSD1306"
```
reel_board uses ssd16xx driver, `SSD16XX`

hakehuang replied:
ok

hakehuang replied:
ok, thanks

vanwinkeljan commented on boards/posix/native_posix/Kconfig.defconfig:
```diff
@@ -80,6 +80,13 @@ config SDL_DISPLAY
+if CHARACTER_FRAMEBUFFER
+
+config CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME
+	default "SSD1306"
```
native_posix board uses sdl driver, `SDL_DISPLAY`

hakehuang replied:
ok

hakehuang replied:
ok, thanks

pabigot commented on subsys/fb/cfb.c:
```diff
@@ -103,6 +103,46 @@ static u8_t draw_char_vtmono(const struct char_framebuffer *fb,
+			}
+			ppixel = (u16_t *)&fb->buf[2 * (fb_y + x + g_x)];
+			v = (glyph_ptr[g_y] & masker) == 0 ? 0 : 255;
+			*ppixel =  ((v>>3)<<11) | ((v>>2)<<5) | (v>>3);
```
This seems a little complex.  Isn't it equivalent to:
```
*ppixel = ((glyph_ptr[g_y] & masker) == 0) ? 0 : 0xFFFF;
```
?

hakehuang replied:
no, this is to convert the 8x8 mono to rgb565 format, here I use v to be r(5bit )g(6 bit)b(5bit).

pabigot replied:
The right hand side converts to either 0 or 0xFFFF depending on whether the masker bit is set.  It's too complex.  Add a comment indicating that 0xFFFF means white in RGB565 encoding, if you want, but don't make maintainers try to figure out what that expression means.

hakehuang replied:
good suggestion, add one comment

pabigot commented on subsys/fb/cfb.c:
```diff
@@ -117,13 +157,31 @@ int cfb_print(struct device *dev, char *str, u16_t x, u16_t y)
+			case PIXEL_FORMAT_MONO10:
+			case PIXEL_FORMAT_MONO01:
+				x += fb->kerning +
+				draw_char_vtmono(fb, str[i], x, y);
```
I'd be happy to see this on one line.  If you really want to obey the length limits (they're not mandatory), please indent the second line to make more clear it's a continuation, rather than a mistake separating two statements.  (I also like to put the operator at the start of the second line, but you might get pushback from others on that.)

Also below.

pabigot replied:
FYI: If you run [uncrustify](https://docs.zephyrproject.org/latest/contribute/index.html#uncrustify) on the source file it fixes style issues in a way nobody should complain about.

hakehuang replied:
ok, let me revise

hakehuang replied:
> A couple comments below. Blockers are:
> 
> * Unnecessary `CHARACTER_FRAMEBUFFER_USE_CST_FONTS` Kconfig
> * Overcomplex expression in `draw_char_vtrgb565`
> * Indentation for split statements
> 
> I also need you to document where to get the new font, and how to regenerate the glyphs. I found the font, and used the updated script in #17541, but can't replicate the bit data, which may be because of changes in that PR or because it was generated with some other tool. IMO you don't have to use that script, but you do need to document how you generated the data: if somebody wants to use the font but needs a different size or weight, they'll want this information.
> 
> FWIW I would like to see this merged soon, as I have work in progress that will need to be rebased on (and can take advantage of) these changes.

I download the font from https://www.download-free-fonts.com/ and the license can be fond in the header file of cfb_fonts8x8.c, and I use on line converter to c array, you can find it here
https://littlevgl.com/ttf-font-to-c-array

pabigot replied:
Please put that in a commit message that adds the font.  We can't expect people to track it down in the github review archives.

hakehuang replied:
ok

pabigot reviewed (changes requested):
A couple comments below.  Blockers are:
* ~Unnecessary `CHARACTER_FRAMEBUFFER_USE_CST_FONTS` Kconfig~ addressed
* ~Overcomplex expression in `draw_char_vtrgb565`~ addressed
* ~Indentation for split statements~ addressed

~I also need you to document where to get the new font, and how to regenerate the glyphs.  I found the font, and used the updated script in #17541, but can't replicate the bit data, which may be because of changes in that PR or because it was generated with some other tool.  IMO you don't have to use that script, but you do need to document how you generated the data: if somebody wants to use the font but needs a different size or weight, they'll want this information.~ provided in commit message

FWIW I would like to see this merged soon, as I have work in progress that will need to be rebased on (and can take advantage of) these changes.

jfischer-no reviewed (commented):
not addressed

hakehuang commented on subsys/fb/Kconfig:
```diff
@@ -13,12 +13,30 @@ menuconfig CHARACTER_FRAMEBUFFER
+	help
+	  Use 8x8 fonts.
+
+config CHARACTER_FRAMEBUFFER_USE_CST_FONTS
```
the initial idea is just to leave an entry to those who want to customer, so that they can use this config to add source code.  such as
`zephyr_sources_ifdef(CONFIG_CHARACTER_FRAMEBUFFER_USE_CST_FONTS cfb_cstfonts.c`

pabigot replied:
Do you really expect customers to add code directly into the main tree, rather than in their application code?  If so, then they can also modify `CMakeLists.txt`.  IMO this option really shouldn't be here.

hakehuang replied:
accept, removed.

hakehuang commented on subsys/fb/cfb.c:
```diff
@@ -117,13 +157,31 @@ int cfb_print(struct device *dev, char *str, u16_t x, u16_t y)
 	}
 
-	if ((fb->screen_info & SCREEN_INFO_MONO_VTILED) && !(y % 8)) {
+	if (!(y % 8)) {
```
Thanks, so I leave it as is.

pabigot commented:
> not addressed

@jfischer-phytec-iot Can you identify what was not addressed?  I'd like to see this merged, so I can rebase #17541 if necessary and get that merged too.

The close and reopen about to happen is because the shippable failure should be resolved by ff0b76ea4ac7bc24f2e016a872f501f0e87e1cc9 which was merged directly to master.

pabigot commented:
reopening to see if shippable passes.

jfischer-phytec-iot commented on samples/display/cfb_custom_font/CMakeLists.txt:
```diff
@@ -9,4 +9,8 @@ project(cfb_custom_font)
 
 set(gen_dir ${ZEPHYR_BINARY_DIR}/include/generated/)
+if(CONFIG_DISPLAY_MCUX_ELCDIF)
+generate_cfb_font_for_target(app fonts/8x8text_darkGrayNoShadow.png ${gen_dir}/cfb_font_dice.h 8 8 --first 49 --last 54)
```
Why is 8x8text_darkGrayNoShadow.png inserted here, what advantages does it bring compared to dice.png?

No word about it in commit message. Really, why was it accepted?

hakehuang replied:
the dice png is very urgly to see in 8x8 format. and I find a nice one so I put it here, the license is free font.

jfischer-phytec-iot replied:
To really understand the problem, what is it for a display? Which graphical display can be more limited than an EPD? Where does the limit comes, is the vertical resolution only 8 pixel?

hakehuang replied:
this is typical for a LCD display, and use a small font has many benifits, 
1. this reuqires less storage size.
2. 8 pixel is select because it is easy to get aligned with byte length and height. 
I try to parase the dice 32x32 font, and for 1 the result is:
```
        /* 49 (1) */
        {
                0x00,0x00,0x00,0x00, /* 0000 0000 0000 0000 0000 0000 0000 0000*/
                0xc0,0xff,0xff,0x03, /* 1100 0000 1111 1111 1111 1111 0000 0011*/
                0xf0,0xff,0xff,0x0f, /* 1111 0000 1111 1111 1111 1111 0000 1111*/
                0xf8,0xff,0xff,0x1f, /* 1111 1000 1111 1111 1111 1111 0001 1111*/
                0xfc,0xff,0xff,0x3f, /* 1111 1100 1111 1111 1111 1111 0001 1111*/
                0xfc,0xff,0xff,0x3f, /* 1111 1100 1111 1111 1111 1111 0011 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0x3f,0xfc,0x7f, /* 1111 1110 0011 1111 1111 1100 0111 1111*/
                0xfe,0x1f,0xf8,0x7f, /* 1111 1110 0001 1111 1111 1000 0111 1111*/
                0xfe,0x1f,0xf8,0x7f, /* 1111 1110 0001 1111 1111 1000 0111 1111*/
                0xfe,0x1f,0xf8,0x7f, /* 1111 1110 0001 1111 1111 1000 0111 1111*/
                0xfe,0x1f,0xf8,0x7f, /* 1111 1110 0001 1111 1111 1000 0111 1111*/
                0xfe,0x3f,0xfc,0x7f, /* 1111 1110 0011 1111 1111 1100 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfe,0xff,0xff,0x7f, /* 1111 1110 1111 1111 1111 1111 0111 1111*/
                0xfc,0xff,0xff,0x3f, /* 1111 1100 1111 1111 1111 1111 0011 1111*/
                0xfc,0xff,0xff,0x3f, /* 1111 1100 1111 1111 1111 1111 0011 1111*/
                0xf8,0xff,0xff,0x1f, /* 1111 1000 1111 1111 1111 1111 0001 1111*/
                0xf0,0xff,0xff,0x0f, /* 1111 0000 1111 1111 1111 1111 0000 1111*/
                0xc0,0xff,0xff,0x03,/* 1100 0000 1111 1111 1111 1111 0000 0011*/
                0x00,0x00,0x00,0x00, /* 0000 0000 0000 0000 0000 0000 0000 0000*/
        },
```

hakehuang replied:
@jfischer-phytec-iot please check my reply.

jfischer-phytec-iot commented on subsys/fb/cfb.c:
```diff
@@ -141,11 +198,11 @@ static int cfb_reverse_bytes(const struct char_framebuffer *fb)
 	for (size_t i = 0; i < fb->x_res * fb->y_res / 8U; i++) {
 		fb->buf[i] = (fb->buf[i] & 0xf0) >> 4 |
-			      (fb->buf[i] & 0x0f) << 4;
+			     (fb->buf[i] & 0x0f) << 4;
```
Please make all style fixes in separate commit.

hakehuang replied:
ok

jfischer-phytec-iot commented on subsys/fb/cfb_fonts8x8.c:
```diff
@@ -0,0 +1,1274 @@
+/*
```
I do not understand why you refuse to use `scripts/gen_cfb_font_header.py`, especially with the improvements from @pabigot . It would improve the maintainability and make it easier to go away with the build-in fonts in general and generate the fonts at the build time.

hakehuang replied:
when I put my font there, I am not aware about @pabigot's commit, and pre his comments it is not obligated to use that. and generate fonts at build time is demoed in the cfb_custom_font sample. I put a standard font here just to want easy customer's usage. @jfischer-phytec-iot  so now I must use pabigot's script to generate my font header?

jfischer-phytec-iot replied:
Please take a look at #17541, it makes possible to generate htiled fonts as yours `cfb_fonts8x8.c`

pabigot replied:
@jfischer-phytec-iot It doesn't right now; at your request I removed hpack support because there's no support for it in the CFB driver.  If #17541 gets merged I can add hpack support in another PR, coordinating with material in this PR.

hakehuang replied:
ok

hakehuang commented on boards/arm/frdm_k64f/Kconfig.defconfig:
```diff
@@ -98,6 +98,13 @@ config SPI_0
 
 endif # SPI
 
+if CHARACTER_FRAMEBUFFER
```
@jfischer-phytec-iot  I have to put it here because the CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME is defined in the subsys/fb/Kconfig, and according to the config rules marcos in samples/display/Kconfig can not override those defined in subsys/fb/Kconfig, and this change is requested by @vanwinkeljan. And user can override this in the sample/display/cfb/Kconfig once we have this defined in board default config.  Actually the CFB driver need to find a supported display, and this selection currently is statically controlled by marco, unless we have a new framwork to replace. but this could be in another patch set.

vanwinkeljan replied:
@hakehuang I did not realize that the board had no display but if this the case I agree with @jfischer-phytec-iot that it should not be put in the boards defconfig. You could move it into the board specific Kconfig overlay of the sampels (eg. `samples/display/cfb_shell/boards/frdm_k64f.conf`)

hakehuang replied:
@vanwinkeljan , @jfischer-phytec-iot , can I ask for your help on which marco shall I use, if Ido not create one in the board Kconfig, I need to use a existing one, but none of the display driver export their driver name with a marco.

vanwinkeljan replied:
I would just add `CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="SSD1306"` to `samples/display/cfb_shell/boards/frdm_k64f.conf` for now, in the end it is the DTS overlay (`samples/display/cfb_shell/frdm_k64f.overlay`) in the sample that sets the name.

jfischer-phytec-iot replied:
I suggest to move all `CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="display"` from  _Kconfig.defconfig_ to samples/display/cfb_shell/boards/_board_.conf. It does not really solve the problem, but not have that much impact.

hakehuang replied:
I see. Thanks a lot.

vanwinkeljan replied:
> Can I make this way?

No as stated before this will make it a big mess as for each display you need to add a new default entry. The error message clearly states the issue, the Kconfig symbol `CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME ` has no prompt.

Could you try by adding a prompt (`string  "Character Framebuffer Display Driver Name"`)?

hakehuang replied:
ok, thanks this works

vanwinkeljan reviewed (changes requested):
Missing prompt in Kconfig symbol `CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME`

vanwinkeljan reviewed (changes requested):
Please also add `CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="SDL_DISPLAY"` to `samples/display/cfb/boards/native_posix_64.conf`

hakehuang commented:
> Please also add `CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="SDL_DISPLAY"` to `samples/display/cfb/boards/native_posix_64.conf`

done. Thanks for letting me know the new board

hakehuang commented on subsys/fb/Kconfig:
```diff
@@ -13,28 +13,39 @@ menuconfig CHARACTER_FRAMEBUFFER
 	  Use default fonts.
 
+config CHARACTER_FRAMEBUFFER_USE_8X8_FONTS
+	bool "Use plain 8x8 fonts"
```
ok

hakehuang replied:
I change to "horizontal tiled font", I miss use the phase

hakehuang commented on subsys/fb/cfb_fonts8x8.c:
```diff
@@ -0,0 +1,1274 @@
+
+	},
+	{
+	/*Unicode: U+0021 (!) , Width: 1 */
```
it is '!' Iguess the missing is due the the font issue, as in the demo we only use 1-9, I did not check every font though

hakehuang replied:
I have split the commit as you recommanded

hakehuang replied:
updated

hakehuang commented on subsys/fb/cfb_fonts8x8.c:
```diff
@@ -0,0 +1,1274 @@
+
+	},
+	{
+	/*Unicode: U+0023 (#) , Width: 4 */
```
according to the unicode, yes. I agree this font is not that perfect, but it is much better than the original ones

hakehuang replied:
updated

hakehuang commented on subsys/fb/cfb_fonts8x8.c:
```diff
@@ -0,0 +1,1274 @@
+
+	},
+	{
+	/*Unicode: U+0042 (B) , Width: 4 */
```
looks like the convertion algorithm will miss some strikes in 8x8

hakehuang replied:
updated

hakehuang commented on samples/display/cfb/boards/frdm_k64f.conf:
```diff
@@ -1,4 +1,4 @@
 CONFIG_SSD1306=y
 CONFIG_SSD1306_SH1106_COMPATIBLE=n
-
+CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="SSD1306"
```
ok, use CONFIG_LVGL_DISPLAY_DEV_NAME

hakehuang commented on samples/display/cfb/boards/mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_HEAP_MEM_POOL_SIZE=300000
+CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="ELCDIF_1"
+
```
remove it

hakehuang replied:
remove this file

hakehuang commented on samples/display/cfb/boards/mimxrt1050_evk.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_HEAP_MEM_POOL_SIZE=300000
+CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="ELCDIF_1"
```
ok

hakehuang commented on samples/display/cfb/boards/native_posix_64.conf:
```diff
@@ -1,3 +1,5 @@
 CONFIG_SDL_DISPLAY_X_RES=250
 CONFIG_SDL_DISPLAY_Y_RES=120
+CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="SDL_DISPLAY"
+
```
remove

MaureenHelm commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk_defconfig:
```diff
@@ -15,3 +15,5 @@ CONFIG_CORTEX_M_SYSTICK=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=600000000
 CONFIG_ARM_MPU=y
+CONFIG_HEAP_MEM_POOL_SIZE=300000
+CONFIG_CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME="ELCDIF_1"
```
Please move these to `boards/arm/mimxrt1050_evk/Kconfig.defconfig`

hakehuang replied:
ok.

MaureenHelm commented on subsys/fb/Kconfig:
```diff
@@ -13,28 +13,39 @@ menuconfig CHARACTER_FRAMEBUFFER
 if CHARACTER_FRAMEBUFFER
 
+
+choice
```
If you make this a named choice, you can override defaults in a Kconfig.defconfig file. For example:
https://github.com/zephyrproject-rtos/zephyr/blob/1593de9642fc0e6509b22862b570c5c4d0618818/soc/arm/nxp_imx/rt/Kconfig.soc#L321-L323
https://github.com/zephyrproject-rtos/zephyr/blob/1593de9642fc0e6509b22862b570c5c4d0618818/boards/arm/mimxrt1050_evk/Kconfig.defconfig#L13-L15

hakehuang replied:
ok, I move the choice to Kconfig.defconfig, but keep the default choice in the subsys/fb/kconfig as this is for subsys bind.

MaureenHelm commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -94,4 +94,17 @@ endchoice
 
 endif # LVGL
 
+if DISPLAY_MCUX_ELCDIF
```
This should be `if CHARACTER_FRAMEBUFFER`

hakehuang replied:
updated, Thanks

jfischer-phytec-iot commented on samples/display/cfb_shell/boards/frdm_k64f.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_I2C=y
```
Remove this file, it should not be there.

hakehuang replied:
ok

jfischer-phytec-iot commented:
@hakehuang Can you verify that it build and works?

hakehuang commented:
> @hakehuang Can you verify that it build and works?

@jfischer-phytec-iot 

- sample cfb is ok on frdm_k64f and mimxrt1050_evk, but fails on real_board, which seems not related to my change.

- sample cfb_customer_font works for mimxrt1050_evk, but fails on real_board with the same reason

- sample cfb_shell works for frdmk64f and mimxrt1050_evk, but fails on real_board with the same reason

please see below build log

[logs.txt](https://github.com/zephyrproject-rtos/zephyr/files/3655786/logs.txt)

jfischer-phytec-iot commented:
@hakehuang 
>     * sample cfb is ok on frdm_k64f and mimxrt1050_evk, but fails on real_board, which seems not related to my change.

Latest master works fine on reel board, just verified.

hakehuang commented:
> @hakehuang
> 
> > ```
> > * sample cfb is ok on frdm_k64f and mimxrt1050_evk, but fails on real_board, which seems not related to my change.
> > ```
> 
> Latest master works fine on reel board, just verified. You probably have to call `west update`

@jfischer-phytec-iot thanks a lot. it works at my side also.

hakehuang commented:
@jfischer-phytec-iot do you have time to review this pull request? I have updated all according to your change requests

jfischer-no reviewed (approved):
sorry for the delay. Tested with few EPD and OLED displays, LGTM

MaureenHelm commented:
`samples/display/cfb_custom_font` works fine on the `reel_board`, but doesn't seem to work on `mimxrt1050_evk`:

![cfb_custom_font](https://user-images.githubusercontent.com/6172539/66755592-ea319d00-ee5d-11e9-998d-732e29cce5d1.jpg)

hakehuang commented:
> `samples/display/cfb_custom_font` works fine on the `reel_board`, but doesn't seem to work on `mimxrt1050_evk`:
> 
> ![cfb_custom_font](https://user-images.githubusercontent.com/6172539/66755592-ea319d00-ee5d-11e9-998d-732e29cce5d1.jpg)

Hi @MaureenHelm ,

this is because the font that original used for customer sample is defines as below
generate_cfb_font_for_target(app fonts/dice.png ${gen_dir}/cfb_font_dice.h 8 8 --first 49 --last 54)

as I set the default font support to 8x8 so it is very hard to see, take below as an example

static const u8_t cfb_font_custom_88[6][8] = {
        /* 49 (1) */
        {
                0x00,
                0xc0,
                0xf0,
                0xf8,
                0xfc,
                0xfc,
                0xfe,
                0xfe,
        },

which is 
 00000000
 11000000
 11110000
 11111000
 11111100
 11111100
 11111110
 11111110

looks like the convert script can only translate the image without resize, so use to create an 8x8 dice by myself, but this can not pass the review, so I keep the dice.png for now.

hakehuang committed:
fb: add support for rgb565

add cfb to support rgb565 FMT

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
fb: add font8x8 for RGB565 panel

add RGB565 pnale to support font8x8
unify the DISPLAY_DEV_NAME by display config

fonts from https://www.fontsquirrel.com/fonts/sinkin-sans
convert tool: https://littlevgl.com/ttf-font-to-c-array-old

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
boards: add default config for fb support

add config define in default board settings

Signed-off-by: Hake Huang <hake.huang@nxp.com>

hakehuang committed:
samples: enable RGB565 panel support for CFB

 1. disable return check
 2. add RGB565 panel CFB support in RT1050 board
 3. add RT1050 to sample white list
 4. add DISPLAY_DEV_NAME to each support board

Signed-off-by: Hake Huang <hake.huang@nxp.com>

galak commented on boards/arm/mimxrt1050_evk/Kconfig.defconfig:
```diff
@@ -94,4 +94,18 @@ endchoice
+config  HEAP_MEM_POOL_SIZE
+	default 300000
+
+config CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME
```
We should be get name from DT not from Kconfig like this.

jfischer-phytec-iot replied:
@galak I think not in that case, the problem here is the lacking of display interface registration/discovery. I am on it.

vanwinkeljan replied:
@galak Agree here with @jfischer-phytec-iot currency this the only way of telling  which display should be used by CFB.

hakehuang replied:
so no change need right? @galak

galak reviewed (dismissed):
Need to get CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME from DT and not Kconfig.

hakehuang commented:
> Need to get CHARACTER_FRAMEBUFFER_DISPLAY_DEV_NAME from DT and not Kconfig.

@galak if you have no objection, I am going to resolve your change request as is.

jskarzynski commented:
Hello guys!
What is the status of this pull-request? Will it be marged? I have some extension for CFB which is based on those changes.

vanwinkeljan commented:
@jskarzynski we are in the middle of a code freeze for 2.1 so I would say at earliest next year.

jfischer-phytec-iot commented:
Probably it would be better to move f80d951 and 1647d15 in special PR as we have other RGB565 display drivers in the tree.

hakehuang commented:
> Probably it would be better to move [f80d951](https://github.com/zephyrproject-rtos/zephyr/commit/f80d95153dfff195bdf56f9bd7b538222b26582d) and [1647d15](https://github.com/zephyrproject-rtos/zephyr/commit/1647d15b75c53f342ce4345fde20116dbe8a7995) in special PR as we have other RGB565 display drivers in the tree.

@jfischer-phytec-iot  not sure I unstand your request, do you mean split this pr into 2? or drop the #6cb778ef4202f24ece0ba4dbc745133093f23ebb and #e9c1bfbe287cb0aaf2d013540497e511b9faefa7

jskarzynski commented:
> Probably it would be better to move [f80d951](https://github.com/zephyrproject-rtos/zephyr/commit/f80d95153dfff195bdf56f9bd7b538222b26582d) and [1647d15](https://github.com/zephyrproject-rtos/zephyr/commit/1647d15b75c53f342ce4345fde20116dbe8a7995) in special PR as we have other RGB565 display drivers in the tree.

I vote for this approach. It would be nice to have those two changes merged.

jfischer-phytec-iot commented:
> > Probably it would be better to move [f80d951](https://github.com/zephyrproject-rtos/zephyr/commit/f80d95153dfff195bdf56f9bd7b538222b26582d) and [1647d15](https://github.com/zephyrproject-rtos/zephyr/commit/1647d15b75c53f342ce4345fde20116dbe8a7995) in special PR as we have other RGB565 display drivers in the tree.
> 
> I vote for this approach. It would be nice to have those two changes merged.

I will test it with st7789v shield and probably #22033

aescolar commented:
@hakehuang what is the status of this PR? if it is not relevant anymore, would you be so kind as to close it?

hakehuang commented:
> @hakehuang what is the status of this PR? if it is not relevant anymore, would you be so kind as to close it?

@aescolar , yes, I have another PR
